File: detect-objects.py
'''
Outline:
This file contains a system for detecting
objects based on an array of distance values in
taken in a circle.'''
# import tabulate

class Object:
    '''Very unfinished. Just holds the number of rays wide
    the object is'''

    def __init__(self, start_index, end_index) -> None:
        self.start_index: int = start_index
        self.end_index: int = end_index
        self.width: int = end_index - start_index + 1

    def __repr__(self) -> str:
        return f'Object: Start: {round(self.start_index, 3)}, End: {round(self.end_index, 3)}, Width: {round(self.width, 3)}'


def cleanData(sample: list[int]) -> list[int]:
    from collections import Counter

    items = Counter(sample).items()
    for value, freq in sorted(items, reverse=True):
        if freq / len(sample) > 0.1: # frequent enough to be a true background value
            max_value = value
            break

    return [min(value, max_value) for value in sample]



def getKPointAverages(distance_values: list[int], k: int, background_value: int) -> list[int]:
    '''Returns a list of the k point averages of the given data'''
    # pad data with background values to allow for trailing k point average
    distance_values = [background_value]*(k-1) + distance_values

    return [sum(distance_values[i-k:i]) / k for i in range(k, len(distance_values)+1)]



def displayObjects(objects: list[Object]) -> None:
    '''Lists the objects in tidy rows and columns'''
    print('Index     Start      End        Width')

    for idx, obj in enumerate(objects):
        rounded_start: str = str(round(obj.start_index, 3))
        rounded_end: str = str(round(obj.end_index, 3))
        rounded_width: str = str(round(obj.width, 3))

        print(f'{str(idx):10s}{rounded_start:10s} {rounded_end:10s} {rounded_width:10s}')



def detectObjects(distances: list[int], k=5, print_data=False) -> list[Object]:
    '''Takes an array of distance values and detects significance
    changes in distances to signify objects.'''
    cleaned_dists: list[int] = cleanData(distances)

    background_dist: int = max(cleaned_dists)
    n: int = len(cleaned_dists)
    print(n)

    k_point_averages: list[int] = getKPointAverages(cleaned_dists, k, background_dist)

    # Initalise variables
    previous_average: int = k_point_averages[0]
    percent_change_threshold: int = 0.05 # arbitrary

    objects: list[Object] = []
    recording_object: bool = (previous_average - background_dist) / background_dist < percent_change_threshold
    object_start: int = 0


    for idx, average in enumerate(k_point_averages):
        percent_change: int = (average - previous_average) / previous_average

        if percent_change < -percent_change_threshold and not recording_object: # start recording new object
            object_start = idx
            recording_object = True

        elif percent_change > percent_change_threshold and recording_object: # stop recording object
            normalised_indices = (object_start / n * 360, (idx-1) / n * 360)
            # objects.append(Object(object_start, idx-1))
            objects.append(Object(*normalised_indices))
            recording_object = False

        if print_data:
            print(f'Idx: {idx}, Dist: {cleaned_dists[idx]}, %Change: {percent_change:.3f}, Detecting: {recording_object}')

        previous_average = average

    if recording_object: # add last object if it's cutoff
        objects.append(Object(object_start, idx))

    return objects

        

if __name__ == '__main__':
    from os import system; system('clear')

    # sample = [10]*30 + [2]*10 + [10]*50 + [3]*20 + [10]*100 + [1]*40 + [10]*110
    # sample = [10]*10 + [5]*10 + [10]*10 + [5]*10
    # sample = [10]*6 + [1]*6 + [10]*10

    # k = 5
    # averages = getKPointAverages(sample, 5, max(sample))
    # padded_sample = [max(sample)]*(k-1) + sample
    # print(' '.join(str(i).zfill(4) for i in range(-k+1, len(padded_sample)-k+1)))
    # print(' '.join(str(value).zfill(4) for value in padded_sample))
    # print(' '*4*5 + ' '.join(str(value).zfill(4) for value in averages))
    # print()

    with open('SonarData.csv') as file:
        lines = file.read().splitlines()
    sample = [int(line.split(',')[0]) for line in lines]


    objects = detectObjects(sample, k=5, print_data=False)
    displayObjects(objects)

    # for obj in objects:
    #     print(obj)

File: detect-objects.c
/*
DetectObjects.c

Created by Thomas Mitchell and Ben Ross on 02.01.2022

This file contains a system for detecting objects based 
on an array of distance values in taken in a circle. 
*/

#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>

typedef struct {
	float start;
	float end;
	float width;
} object_t;

void sortDescending(int *array, int n) {
	/* Sort elements of an array in descending order */
	int i, j, tmp;
	
	for (i = 0; i < n; i++) {
        for (j = i+1; j < n; j++) {
            if (array[i] < array[j]) {
                tmp = array[i];
                array[i] = array[j];
                array[j] = tmp;
            }
        }
    }
}

int countFreq(int *sample, int *values, int *freq, int n) {
	/* Count frequency of elements in sample */
	int i, j, count, numUnique=0;
   
	for(i = 0; i < n; i++) {
		/* Copy elements from data to values */
		values[i] = sample[i];
        /* Initially initialize frequencies to -1 */
        freq[i] = -1;
    }
	
	/* Sort elements of an array in descending order */
	sortDescending(values, n);
	 
	/* Traverse through array elements and count frequencies */
    for (i = 0; i < n; i++) {
        count = 1;
        for (j = i+1; j < n; j++)
        {
            /* If duplicate element is found */
            if(values[i] == values[j]) {
				values[i] = values[j];
                count++;

                /* Make sure not to count frequency of same element again */
                freq[j] = 0;
            }
        }

        /* If frequency of current element is not counted */
        if (freq[i] != 0) {
            freq[i] = count;
        }
	}
	
	/* Assign unique elements to beginning of arrays */
	for (i = 0; i < n; i++) {
		if (freq[i] != 0 && values[i] != 0) {
			values[numUnique] = values[i];
			freq[numUnique] = freq[i];
			numUnique++;
		}
	}
	
	return numUnique;
}

int *cleanData(int *sample, int n) {
	/* Remove noise from sample */ 
	int *cleanData = calloc(n, sizeof(int));
	int *values = calloc(n, sizeof(int));
	int *freq = calloc(n, sizeof(int));
	int maxValue, numUnique;
	
	/* Get frequency of sample elements */ 
	numUnique = countFreq(sample, values, freq, n);

	/* Find background value */ 
	for (int i = 0; i < numUnique; i++) {
		/* If frequent enough to be a true background value */
		if ((float) freq[i]/n > 0.1) { 
			maxValue = values[i];
			break;
		}
	}
	
	/* Remove noise from sample by overwriting noise with backgroud value */ 
	for (int i = 0; i < n; i++) {
		if (maxValue < sample[i]) {
			cleanData[i] = maxValue;
		} else {
			cleanData[i] = sample[i];
		}
	}
	
	/* Free dynamically allocated arrays from memory */
	free(values);
	free(freq);
	
	return cleanData;
}

float *getKPointAverages(int *distVals, int backgroundDist, int k, int n2) {
	/* Returns a list of the k point averages of the given data */
	int n1 = k-1;
	int n3 = n1 + n2;
	int sum=0, pos=0;

	float *kPointAverages = calloc(n3,sizeof(int));
	int *paddedDistVals = calloc(n3, sizeof(int));
	int *arrayNumbers = calloc(k, sizeof(int));
	
	/* Pad data with background values to allow for trailing k point average */
	for (int i = 0; i < n1; i++) {
		paddedDistVals[i] = backgroundDist;
	}
	
	/* Copy elements from distVals to paddedDistVals */
	for (int i = 0, j = n1; j < n3; i++, j++) {
		paddedDistVals[j] = distVals[i];
	}
	
	/* Calculate moving average */
	for (int i = 0; i < n3; i++) {
		/* Subtract the oldest number from the sum, add the new number */
		sum = sum - arrayNumbers[pos] + paddedDistVals[i];
		/* Assign the nextNum to the position in the array */
		arrayNumbers[pos] = paddedDistVals[i];
		
		if (i < k) {
			/* First k avergaes */
			kPointAverages[i] = (float) sum/(i+1);
		} else {
			kPointAverages[i] = (float) sum/k;
		}
		
		pos++;
		if (pos >= k) {
			pos = 0;
		}
	}
	
	/* Free arrays from memory */ 
	free(paddedDistVals);
	free(arrayNumbers);
	
	return kPointAverages;
}

void displayObjects(object_t *objects, int n) {
	/* Lists the objects in tidy rows and columns */
	float start, end, width;
	
	printf("Index     Start       End    Width\n");

	for (int i = 0; i < n; i++) {
		object_t obj = objects[i];
		start = obj.start;
		end = obj.end;
		width = obj.width;
		printf("%d %13.3f %9.3f %8.3f\n", i, start, end, width);
	}
}

int max(int *array, int n) {
	/* Returns max value in array */
	int tmp = array[0];
	for (int i = 1; i < n; i++) {
		if (array[i] > tmp) {
			tmp = array[i];
		}
	}
	return tmp;
}

void setWidth(object_t *obj) {
	/* Sets width of object */ 
	(*obj).width = (*obj).end-(*obj).start+1;
}

float normalise(int num, int n) {
	return (float) num * 360 / n;
}
	

void *detectObjects(int *sample, int n, int *num, int k, bool printData) {
	/* Takes an array of distance values and detects significance
    changes in distances to signify objects */	
	
	float percentChange, percentChangeThreshold = 0.05; /* Arbitrary */
	float previousAverage, *kPointAverages;
	
	int *cleanDists, backgroundDist, numAverages, idx, objectStart = 0;
	
	object_t *objects = calloc(10, sizeof(object_t)); /* Max 10 objects */
	
	bool recordingObject = false;

	/* Remove noise from sample */
	cleanDists = cleanData(sample, n); 
	/* Find true distance value */
	backgroundDist = max(cleanDists, n); 
	/* Get moving averages */
	kPointAverages = getKPointAverages(cleanDists, backgroundDist, k, n);
	
	/* Number of averages due to additional padded data */
	numAverages = n+k-1;
	previousAverage = kPointAverages[0];
	
	/* If object is being detected */ 
	recordingObject = (previousAverage-backgroundDist)/backgroundDist<percentChangeThreshold;
	
	for (idx = 0; idx < numAverages; idx++) {
		/* Find degree of change between previous average */ 
		percentChange = (kPointAverages[idx]-previousAverage)/previousAverage;
		
		/* If recording object */
		if (percentChange < -percentChangeThreshold && !recordingObject) {
			objectStart = idx;
			recordingObject = true;
		} /* If trailing edge of object detected */
		else if (percentChange > percentChangeThreshold && recordingObject) {
			objects[*num].start = normalise(objectStart, n);
			objects[*num].end = normalise(idx-1, n);
			setWidth(objects+(*num));
			recordingObject = false;
			(*num)++;
		}
			
		if (printData) {
			printf("Idx: %d, ", idx);
			printf("Dist: %d, ", cleanDists[idx]);
			printf("%%Change: %f, ", percentChange);
			printf("Detecting: %s\n, ", recordingObject ? "true" : "false");
		}
		
		previousAverage=kPointAverages[idx];
	}
	
	if (recordingObject) { /* Add last object if it's cutoff */
		if (objects[0].start < 5) { // wrap around 
			objects[0].start = normalise(idx-1, n);
			// take off extra width
		} else {
			objects[*num].start = normalise(objectStart, n);
			objects[*num].end = normalise(idx-1, n);
		}
	}
	
	/* Free array from memeory */
	free(kPointAverages);
	
	return objects;	
}


int main () {
	FILE* file = fopen("SonarData.csv", "r");
	int sampleLength = 1;
	int numObjects = 0;
	int sample[2048];
	int k = 5;
	
	/* Write data from file to array */
	fscanf (file, "%d\n", sample);    
	while (!feof (file)) { /* While not end of file */
		fscanf (file, "%d\n", (sample+sampleLength)); 
		sampleLength++;
    }
	fclose(file); /* Close file */  

	/* Analyse data */
	object_t *objects = detectObjects(sample, sampleLength, &numObjects, k, true);
	displayObjects(objects, numObjects); /* Display objects */
	free(objects);
	return 0;
}